### PR TITLE
feat: improve build images

### PR DESCRIPTION
**Pull request by hoffa**
#### Which issue(s) does this change fix?

N/A

#### Why is this change necessary?

Similar to #2560, but only makes changes not affecting base images.

#### How does it address the issue?

- Use fewer layers where sensible (shaves off about 50 MB from every image).
- Fix Python 2.7 build image.
- Fail build script if any build fails.
- Remove `curl -k`s.

#### What side effects does this change have?

None.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: build-image-src/Dockerfile-python27 [hunk 1]
```diff
@@ -56,6 +55,6 @@ ENV LANG=en_US.UTF-8
 # Python for it to be picked up during `sam build`
 RUN pip3 install wheel
 
-RUN curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python get-pip.py && rm get-pip.py
```
Comment (awood45) [inline build-image-src/Dockerfile-python27:59]: Has this been tested today? Making sure this isn't impacted by the recent `pip` deprecation (which frankly might be a broader issue for building Python 2.7 going forward beyond our scope).

**Review by awood45 — APPROVED** [review]
This generally looks great! One testing question

File: build-image-src/Dockerfile-java8-al2 [hunk 2]
```diff
@@ -5,8 +5,8 @@ FROM amazonlinux:2
 COPY --from=0 / /rootfs
 
 # Installing by yum at copied location
-RUN yum groupinstall -y development --installroot=/rootfs && yum clean all
```
Comment (sriram-mv) [inline build-image-src/Dockerfile-java8-al2:8]: why are we removing `yum clean all` ? is it because there is not any cache to clear?

**Review by sriram-mv — APPROVED** [review]
Looks good to me, I had one question.

File: build-image-src/Dockerfile-python27 (hunk 1 above)
Comment (hoffa) [reply build-image-src/Dockerfile-python27:59]: I've tested it builds, `pip` links to the Python 2 `pip`, and that a dummy `pip install` works fine.

File: build-image-src/Dockerfile-java8-al2 (hunk 2 above)
Comment (hoffa) [reply build-image-src/Dockerfile-java8-al2:8]: It's redundant as it's run L27 in the same layer.

File: build-image-src/Dockerfile-python27 (hunk 1 above)
Comment (hoffa) [reply build-image-src/Dockerfile-python27:59]: And for reference, that was added fairly recently: https://github.com/pypa/get-pip/pull/81

The homepage hasn't yet been updated, but refers to the same method for Python 2.6: https://pip.pypa.io/en/stable/installing/#installing-with-get-pip-py

Also https://stackoverflow.com/a/65866547

**Review by jfuss — APPROVED** [review]
LGTM

**Pull request merged**